build/i2c/src/lib.rs

```diff
@@ -761,6 +767,118 @@ impl ConfigGenerator {
         Ok(())
     }
 
+    pub fn generate_validation(&mut self) -> Result<()> {
+        write!(
+            &mut self.output,
+            r##"
+    pub mod validation {{
+        use drv_i2c_api::{{I2cDevice, Controller, PortIndex}};
+        use drv_i2c_devices::Validate;
+        use userlib::TaskId;
+
+        pub enum I2cValidation {{
+            RawReadOk,
+            Good,
+            Bad,
+        }}
+
+"##
+        )?;
+
+        //
+        // Lord, have mercy: we are going to find the crate containing i2c
+        // devices, and go fishing for where we believe the device drivers
+        // themselves to be.  It does not need to be said that this is
+        // operating by convention; there are (many) ways to envision this
+        // breaking -- with apologies, dear reader, if that's what brings you
+        // here!
+        //
+        use cargo_metadata::MetadataCommand;
+
+        let metadata = MetadataCommand::new()
+            .manifest_path("./Cargo.toml")
+            .exec()
+            .unwrap();
+
+        let pkg = metadata
+            .packages
+            .iter()
+            .find(|p| p.name == "drv-i2c-devices")
+            .context("failed to find drv-i2c-devices")?;
+
+        let dir = pkg
+            .manifest_path
+            .parent()
+            .context("failed to get i2c device path")?;
+
+        let mut drivers = std::collections::HashSet::new();
+
+        println!("cargo:rerun-if-changed={}", dir.join("src").display());
+
+        for entry in std::fs::read_dir(dir.join("src"))? {
+            if let Some(f) = entry?.path().file_name() {
+                if let Some(s) = f.to_str() {
+                    if let Some(name) = s.strip_suffix(&".rs") {
+                        drivers.insert(name.to_string());
+                    }
+                }
```

Comment:
We should probably `unwrap` the `OsStr` to `&str` conversion, since if that fails, something is very weird in the file list.

(also, you don't need `&` in front of the string)

```suggestion
                if let Some(name) = f.to_str().unwrap().strip_suffix(".rs") {
                    drivers.insert(name.to_string());
                }
```

Reply:
Fixed